From: Suresh Siddha <suresh.b.siddha@intel.com>
Subject: x86, apic: Don't use logical-flat mode when CPU hotplug may exceed 8 CPUs
References: bnc#581567
Patch-Mainline: yes (2.6.33-rc8)
Commit-ID: 681ee44d40d7c93b42118320e4620d07d8704fd6

Signed-off-by: Thomas Renninger <trenn@suse.de>

We need to fall back from logical-flat APIC mode to physical-flat mode
when we have more than 8 CPUs.  However, in the presence of CPU
hotplug(with bios listing not enabled but possible cpus as disabled cpus in
MADT), we have to consider the number of possible CPUs rather than
the number of current CPUs; otherwise we may cross the 8-CPU boundary
when CPUs are added later.

32bit apic code can use more cleanups (like the removal of vendor checks in
32bit default_setup_apic_routing()) and more unifications with 64bit code.
Yinghai has some patches in works already. This patch addresses the boot issue
that is reported in the virtualization guest context.

[ hpa: incorporated function annotation feedback from Yinghai Lu ]

Signed-off-by: Suresh Siddha <suresh.b.siddha@intel.com>
LKML-Reference: <1265767304.2833.19.camel@sbs-t61.sc.intel.com>
Acked-by: Shaohui Zheng <shaohui.zheng@intel.com>
Reviewed-by: Yinghai Lu <yinghai@kernel.org>
Cc: <stable@kernel.org>
Signed-off-by: H. Peter Anvin <hpa@zytor.com>

Automatically created from "patches.arch/x86_64_apic_consider_hotplug_for_mode_logical_flat.patch" by xen-port-patches.py

--- sle11sp1-2010-03-01.orig/arch/x86/kernel/mpparse-xen.c	2009-11-06 10:52:22.000000000 +0100
+++ sle11sp1-2010-03-01/arch/x86/kernel/mpparse-xen.c	2010-03-01 14:45:20.000000000 +0100
@@ -371,13 +371,6 @@ static int __init smp_read_mpc(struct mp
 		x86_init.mpparse.mpc_record(1);
 	}
 
-#ifdef CONFIG_X86_BIGSMP
-	generic_bigsmp_probe();
-#endif
-
-	if (apic->setup_apic_routing)
-		apic->setup_apic_routing();
-
 	if (!num_processors)
 		printk(KERN_ERR "MPTABLE: no processors registered!\n");
 	return num_processors;
